lib/elixir/lib/module/types/pattern.ex

```diff
@@ -214,71 +217,88 @@ defmodule Module.Types.Pattern do
+    # TODO: Handle no matching signature
+    signature = guard_signature(guard, length(args))
 
     # Only check type guards in the context of and/or/not,
     # a type guard in the context of is_tuple(x) > :foo
     # should not affect the inference of x
     if not type_guard? or consider_type_guards? do
+      stack = push_expr_stack(expr, stack)
+      param_unions = signature_to_param_unions(signature, context)
       arg_stack = %{stack | type_guards: {false, keep_guarded?}}
+      mfa = {:erlang, guard, length(args)}
 
       with {:ok, arg_types, context} <-
-             map_reduce_ok(args, context, &of_guard(&1, arg_stack, &2)),
-           {:ok, context} <- unify_call(arg_types, param_types, stack, context) do
-        {arg_types, guard_sources} =
-          case arg_types do
-            [{:var, index} | rest_arg_types] when type_guard? ->
-              guard_sources = Map.put_new(context.guard_sources, index, :guarded)
-              {rest_arg_types, guard_sources}
-
-            _ ->
-              {arg_types, context.guard_sources}
-          end
-
-        guard_sources =
-          Enum.reduce(arg_types, guard_sources, fn
-            {:var, index}, guard_sources ->
-              Map.update(guard_sources, index, :fail, &guarded_if_keep_guarded(&1, keep_guarded?))
-
-            _, guard_sources ->
-              guard_sources
-          end)
-
+             map_reduce_ok(Enum.zip(args, param_unions), context, fn {arg, param}, context ->
+               of_guard(arg, param, arg_stack, context)
+             end),
+           {:ok, return_type, context} <-
+             unify_call(arg_types, signature, expected, mfa, stack, context, type_guard?) do
+        guard_sources = guard_sources(arg_types, type_guard?, keep_guarded?, context)
         {:ok, return_type, %{context | guard_sources: guard_sources}}
       end
     else
+      # Assume type guard
+      [{_params, return_type}] = signature
       {:ok, return_type, context}
     end
   end
 
   # map.field
-  def of_guard({{:., meta1, [map, field]}, meta2, []}, stack, context) do
-    of_guard({{:., meta1, [:erlang, :map_get]}, meta2, [field, map]}, stack, context)
+  def of_guard({{:., meta1, [map, field]}, meta2, []}, expected, stack, context) do
+    of_guard({{:., meta1, [:erlang, :map_get]}, meta2, [field, map]}, expected, stack, context)
   end
 
   # var
-  def of_guard(var, _stack, context) when is_var(var) do
+  def of_guard(var, _expected, _stack, context) when is_var(var) do
     {:ok, get_var!(var, context), context}
   end
 
-  def of_guard(expr, stack, context) do
-    of_shared(expr, stack, context, &of_guard/3)
+  def of_guard(expr, _expected, stack, context) do
+    of_shared(expr, stack, context, &of_guard(&1, :dynamic, &2, &3))
+  end
+
+  defp signature_to_param_unions(signature, context) do
+    signature
+    |> Enum.map(fn {params, _return} -> params end)
+    |> zip_many()
+    |> Enum.map(&to_union(&1, context))
```

Comment:
It would be good if we can avoid multiple passes. We should either have a specific zip_many for this case, that does the pass zips and extracts OR generalize zip_many to unpack/pack as it goes. I think the former is simpler.

Reply:
I think the complexity of `zip_many` would increase too much and it's already quite complex. I managed to remove one pass from the `zip_many` implementation though. Feel free to take a look at it.